### PR TITLE
add with_timezone

### LINKED ISSUE #5905: Implement `with_timezone` function
### Is your feature request related to a problem or challenge?

per discussion in https://github.com/apache/arrow-rs/issues/1936#issuecomment-1309690020
casting now consider timezone

i'd like a new function `with_timezone` that could simply change the timezone in datatype without changing any underline value

### Describe the solution you'd like

implement a new function `with_timezone(Timestamp, 'Some-TimeZone')`

e.g.
`with_timezone(timestamp '2000-01-01T00:00:00', '+08:00')` becomes `2000-01-01T08:00:00+08:00`
`with_timezone(with_timezone(timestamp '2000-01-01T00:00:00', '+08:00'), '+07:00')` becomes `2000-01-01T07:00:00+07:00`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by waitingkuo**
# Which issue does this PR close?

Closes #5905 

# Rationale for this change

# What changes are included in this PR?

a new function `with_timezone` that replace the timezone of timestamp without changing underline value

```bash
❯ select WITH_TIMEZONE(timestamp '2000-01-01T00:00:00.123456789', '+08:00');
+--------------------------------------------------------------------+
| withtimezone(Utf8("2000-01-01T00:00:00.123456789"),Utf8("+08:00")) |
+--------------------------------------------------------------------+
| 2000-01-01T08:00:00.123456789+08:00                                |
+--------------------------------------------------------------------+
1 row in set. Query took 0.004 seconds.
```

```bash
❯ select WITH_TIMEZONE(WITH_TIMEZONE(timestamp '2000-01-01T00:00:00.123456789', '+08:00'), '+07:00');
+-------------------------------------------------------------------------------------------------+
| withtimezone(withtimezone(Utf8("2000-01-01T00:00:00.123456789"),Utf8("+08:00")),Utf8("+07:00")) |
+-------------------------------------------------------------------------------------------------+
| 2000-01-01T07:00:00.123456789+07:00                                                             |
+-------------------------------------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

while the timezone is incorrect, it raises error
```bash
❯ select WITH_TIMEZONE(timestamp '2000-01-01T00:00:00.123456789', '+25:00');
Arrow error: Parser error: Invalid timezone "+25:00": '+25:00' is not a valid timezone
```

# Are these changes tested?

yes, test cases are added to `timestamps.slt`

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/function.rs [hunk 1]
```diff
@@ -262,6 +262,25 @@ pub fn return_type(
...
+                Ok(DataType::Timestamp(TimeUnit::Millisecond, None))
+            }
+            DataType::Timestamp(TimeUnit::Second, _) => {
+                Ok(DataType::Timestamp(TimeUnit::Second, None))
+            }
+            _ => return Err(DataFusionError::Internal(
+                "The with_timezone function can only accept timestamp as the first arg."
+                    .to_string(),
+            )),
+        },
```
Comment (waitingkuo) [inline datafusion/expr/src/function.rs:282]:
```rust
pub fn return_type(
    fun: &BuiltinScalarFunction,
    input_expr_types: &[DataType],
) -> Result<DataType> {
```
this function only has 2 inputs: function pointer and args types. there's no way to inference timezone in the final return type

File: datafusion/physical-expr/src/functions.rs [hunk 2]
```diff
@@ -169,6 +169,22 @@ pub fn create_physical_expr(
...
+            let timezone = format!("{}", input_phy_exprs[1]);
+            data_type = match data_type {
+                DataType::Timestamp(TimeUnit::Nanosecond, _) => DataType::Timestamp(TimeUnit::Nanosecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Microsecond, _) => DataType::Timestamp(TimeUnit::Microsecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Millisecond, _) => DataType::Timestamp(TimeUnit::Millisecond, Some(timezone)),
+                DataType::Timestamp(TimeUnit::Second, _) => DataType::Timestamp(TimeUnit::Second, Some(timezone)),
+                other => return Err(DataFusionError::Internal(format!(
+                    "Unsupported data type {other:?} as the first arg for function with_timezone",
+                )))
+            };
```
Comment (waitingkuo) [inline datafusion/physical-expr/src/functions.rs:185]: correct the timezone in return data_type according to the args value

@alamb is this implementation ok for now? or it's recommended to modify the `return_type` function to accept some new args to achieve this?

I could submit another ticket to modify it as well

**Review by waitingkuo — COMMENTED** [review]
since there's no way  to extract timezone`fn return_type` , the return `data_type` is  corrected in `create_physical_expr`
@alamb please help to evaluate whether it make sense

cc @tustvold

**Review by alamb — COMMENTED** [review]
Thanks @waitingkuo  -- 

Even though `arrow_cast` could (and probably should) be updated to support timezones having a specific WITH_TIMEZONE seems like a good idea to me as it will be a better experience

I think this PR needs
1. Get the type correct (I left a comment about how to do this)
2. Add an entry to `WITH_TIMEZONE` the user documentation (e.g. https://github.com/apache/arrow-datafusion/blob/f00ef9d5484039dc7013af5b8fa04f7b29ffaba7/docs/source/user-guide/sql/scalar_functions.md?plain=1#L1208)

**Comment by waitingkuo** [comment]
> Thanks @waitingkuo --
> 
> Even though `arrow_cast` could (and probably should) be updated to support timezones having a specific WITH_TIMEZONE seems like a good idea to me as it will be a better experience
> 
> I think this PR needs
> 
> 1. Get the type correct (I left a comment about how to do this)
> 2. Add an entry to `WITH_TIMEZONE` the user documentation (e.g. https://github.com/apache/arrow-datafusion/blob/f00ef9d5484039dc7013af5b8fa04f7b29ffaba7/docs/source/user-guide/sql/scalar_functions.md?plain=1#L1208
>    )

Hi @alamb thanks, in fact i'm working on adding timezone support for `arrow_cast` now, will send a pr soon

**Comment by tustvold** [comment]
> In fact i'm working on adding timezone support for arrow_cast now, will send a pr soon

I think we may want to fix https://github.com/apache/arrow-rs/issues/1936 upstream, before exposing the currently incorrect timezone casting behaviour in arrow-rs?

**Comment by waitingkuo** [comment]
> Thanks @waitingkuo --
> 
> Even though `arrow_cast` could (and probably should) be updated to support timezones having a specific WITH_TIMEZONE seems like a good idea to me as it will be a better experience
> 
> I think this PR needs
> 
> 1. Get the type correct (I left a comment about how to do this)
> 2. Add an entry to `WITH_TIMEZONE` the user documentation (e.g. https://github.com/apache/arrow-datafusion/blob/f00ef9d5484039dc7013af5b8fa04f7b29ffaba7/docs/source/user-guide/sql/scalar_functions.md?plain=1#L1208
>    )

Hi @alamb  thank you for the suggestion.

does it mean that i need to move `WITH_TIMEZONE` out of `BuiltinScalarFunction` ?

**Comment by alamb** [comment]
> does it mean that i need to move WITH_TIMEZONE out of BuiltinScalarFunction ?

I am not sure to be honest -- there might not be a good existing pattern to follow for such functions. :tin

**Comment by waitingkuo** [comment]
@alamb thanks, i'll think about how to fix this

**Comment by alamb** [comment]
Since this has been open for more than a year, closing it down. Feel free to reopen if/when you keep working on it.

**Pull request closed without merging**